src/components/Experiments/ExperimentEditModal.js

```diff
@@ -0,0 +1,195 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+
+import {
+  Button,
+  Modal,
+  FormGroup,
+  FormControl,
+  Glyphicon,
+  Row,
+  Col,
+} from 'react-bootstrap';
+
+import {
+  showEditModal,
+  create as experimentCreate,
+  update as experimentUpdate,
+  uploadFilePairs,
+  UPLOAD_RES_SUCCESS,
+  UPLOAD_RES_FAILURE,
+} from '../../state/experiments';
+
+import './ExperimentEditModal.less';
+
+class ExperimentEditModal extends Component {
+  constructor(props) {
+    super(props);
+
+    const exp = this.props.editModalExperiment;
+    if (exp !== undefined) {
+      this.state = { nameVal: exp.name, descriptionVal: exp.description };
+    } else {
+      this.state = {
+        nameVal: '',
+        descriptionVal: '',
+      };
+    }
+
+    this.handleClose = this.handleClose.bind(this);
+    this.handleNameChange = this.handleNameChange.bind(this);
+    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
+    this.handleSave = this.handleSave.bind(this);
+    this.handleUpload = this.handleUpload.bind(this);
+  }
+
+  handleClose() {
+    this.props.showEditModal(false);
+  }
+
+  handleNameChange(e) {
+    this.setState({ nameVal: e.target.value });
+  }
+
+  handleDescriptionChange(e) {
+    this.setState({ descriptionVal: e.target.value });
+  }
+
+  handleSave() {
+    const create = this.props.editModalExperiment === undefined;
+
+    if (create) {
+      this.props.experimentCreate(
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    } else {
+      this.props.experimentUpdate(
+        this.props.editModalExperiment.id,
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    }
+  }
+
+  handleUpload(e) {
+    this.props.uploadFilePairs(
+      this.props.editModalExperiment.id,
+      e.target.files[0]
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const exp = nextProps.editModalExperiment;
```

Comment:
I see this code 2 times. The copy-past itself is not a problem. But component depends on both constructor and componentWillReceiveProps. So any the change must be done in both places. It would be much better to extract this code to function which returns state. And do `this.state = fn(props)` in constructor and `this.setState(fn(nextProps))` here.

Reply:
Makes sense, it's an easy improvement. Changed in https://github.com/src-d/code-annotation/pull/195/commits/dd9e4fb3943b1f65200e19dec7b9e6ce580973d7